### ISSUE TITLE
Trainer improvements

### ISSUE THREAD

**Issue by jscanvic** [deepinv/deepinv#553]
Here's a few improvements I think would be useful to make:

* [x] Saving the eval loss #777
* [ ] Saving the actual best checkpoint (it's not actually the best when resuming training, it's only the best for the current training segment)
* [x] Make it so we can use the eval loss as criterion for the best checkpoint #777
* [ ] Make it possible to save what's saved to wandb on disk as well
* [x] Support MLflow in addition to wandb #668
* [ ] Make the trainer easier to work with for a train/val/test workflow instead of the current train/val-test workflow (progress in #777 )
* [ ] Improve the post-training (loading the best weights automatically instead of using the last weights)
* [ ] Log the learning rate used at every epoch
* [x] "No learning" metrics should be computed by passing the "no learning" reconstructor instead of the trained one #788
* [ ] Use 1-based indexing for reporting (Epochs 1 to 500, not 0 to 499)
* [ ] Save the path to the run directory in mlops tools
* [ ] Save the path to the best checkpoint in mlops tools
* [ ] Log test images to mlops tools
* [ ] Save the last checkpoint in addition to/instead of historical checkpoints
* [ ] Add an option to disable periodic checkpointing
* [ ] Add support for mixed-precision training #1205

**Dubious code**

1. [x] #904

https://github.com/deepinv/deepinv/blob/ce2b1a33f3afc72f9ee0b59375b57a74d5928e9f/deepinv/training/trainer.py#L903-L913

* It iterates on the batch dimension but does basically the same thing every iteration.
* It appears to save batched images despite iterating on the batch dimension (intended?)
* It increments the counter in the inner loop but probably shouldn't.
